keps/sig-api-machinery/20190206-watch-bookmark.md

```diff
@@ -0,0 +1,184 @@
+    Type EventType
+    Object runtime.Object
+  }
+```
+
+As a result, we will represent bookmark event by setting Bookmark type and
+Object of appropriate type with just ObjectMeta.ResourceVersion field set.
+
+Unfortunately, such a change would break existing clients, as an example
+[official decoder][]. As a result, we will extend `ListOptions` (which is
+how we pass options to watch) with a boolean field where user can opt-in
+for watch bookmarks:
+```
+  type ListOptions struct {
+    ...
+    RequestBookmarks bool
+  }
+```
+We consciously make it just a boolean flag - this gives kube-apiserver an
+ability to choose when they should be send without setting any expectations
+on user side how frequently and when it would be happening. In particular
+client isn't guaranteed to get any bookmarks.
+Such addition to `ListOptions` is also safe from backward-compatibility
+point of view, old kube-apiserver will simply drop this field if set.
+
+Once the API is extended, we add a support for sending bookmarks to watchcache.
+The exact policy of sending them is to be determined, the ideas include:
+- send a bookmark if there weren't any event send to user in the last X
+seconds
+- if we know we will be closing (e.g. due to timeout) a watch, try to send
+a bookmark immediately before that
+
+Finalizing the decision shouldn't block the initial version of this KEP.
+
+[official decoder]: https://github.com/kubernetes/kubernetes/blob/5d4795e14e02ac29273009d86ba3c5012684d5f4/staging/src/k8s.io/client-go/rest/watch/decoder.go#L57
+
+
+### Risks and Mitigations
+
+Sending "watch bookmarks" may break clients not understanding them.
+As a result, we make them explicitly opt-in.
+
+## Graduation Criteria
+
+Beta:
+- Proved scalability/performance gain. With a simple POC I was able to
+reduce amount of processed "init events" by ~40x. So setting the minimum
+goal on 10x without adding any additional visible overhead.
+- Generated informers make use of this new API.
+
+GA:
+- Enabled by default in Kubelet for watching pods for a release.
+- No complaints about the API for a release
+
+## Implementation History
+
+- RRRR-MM-DD: KEP Summary, Moativation and Proposal merged
+
+## Alternatives
+
+The most important considered alternatives are mentioned below.
+
+### Cache in kube-apiserver
+
+Instead of introducing an API for bookmarks, we can try memorizing in watchcache
+what we already processed for a watcher and when it is restarted use that
+information. However, that would require being able to identify and match a
+watcher across restarts which is non-trivial. Moreover, it doesn't work in
+HA setups with multiple kube-apiserver. The idea was dropped.
+
+### API for send bookmark
+
+This is similar to what was done an etcd, where an API was added to notify
+all watchers about current resourceVersion. However, such API would be hard
+to manage in kube-apiserver, and we don't really want to notify everyone at
+the same time. The idea was dropped.
```

Comment:
nit: you could call the section "rejected alternatives" and then you wouldn't have to end each section with "the idea was dropped" :)

Reply:
done